### PR TITLE
Add advanced startup example

### PR DESCRIPTION
**Pull request by mikeshardmind**
## Summary

shows setup_hook and async context manager use with examples
approximating real world use cases

## Checklist

- [X] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [X] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: examples/advanced_startup.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+async def main():
+
+    # When taking over how the bot process is run, you become responsible for a few additional things.
+
+    # 1. logging
+
+    # for this example, we're just going to use some basic defaults provided by python
+    # for more info on setting up logging, see https://docs.python.org/3/howto/logging.html
+    logging.basicConfig(level=logging.INFO)
```
Comment (Rapptz) [inline examples/advanced_startup.py:57]: d.py has a dedicated logging page which is decently in-depth for those who just want quick examples, might want to refer to it too and use one of the snippets.

File: examples/advanced_startup.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import discord
+from discord.ext import commands
+from aiohttp import ClientSession
+
+
+class CustomBot(commands.Bot):
+    def __init__(
+        self,
+        *args,
+        initial_extensions: list,
```
Comment (Rapptz) [inline examples/advanced_startup.py:19]:
```suggestion
        initial_extensions: List[str],
```

File: examples/advanced_startup.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
+# This example covers advanced startup options and uses some real world examples for why you may need them.
+
+import asyncio
+import logging
+import os
+
+from typing import Optional
```
Comment (Rapptz) [inline examples/advanced_startup.py:7]:
```suggestion
from typing import Optional, List
```

**Review by Rapptz — CHANGES_REQUESTED** [review]

File: examples/advanced_startup.py (hunk 1 above)
Comment (mikeshardmind) [reply examples/advanced_startup.py]: Sure, this is a good place to advocate for devs to do more with logging than the basics anyway once they step into needing more fine-grained control over start.

**Review by Rapptz — APPROVED** [review]
Seems fine, thanks.

**Pull request merged**